src/server/pps/server/master.go

```diff
@@ -42,8 +44,31 @@ var (
 	falseVal bool  // used to delete RCs in deletePipelineResources and restartPipeline()
 )
 
+type pipelineKey string
+
+func toKey(p *pps.Pipeline) pipelineKey {
+	return pipelineKey(fmt.Sprintf("%s/%s", p.GetProject().GetName(), p.GetName()))
```

Comment:
I think the null pointer here from p.Project is desirable since it's a problem to pass a projectless pipeline here